Content.Server/Power/Components/ApcComponent.cs

```diff
@@ -1,211 +1,54 @@
-                _uiDirty = true;
-            }
-
-            var extPowerState = CalcExtPowerState();
-            if (extPowerState != _lastExternalPowerState && _lastExternalPowerStateChange + TimeSpan.FromSeconds(VisualsChangeDelay) < _gameTiming.CurTime)
-            {
-                _lastExternalPowerState = extPowerState;
-                _lastExternalPowerStateChange = _gameTiming.CurTime;
-                _uiDirty = true;
-            }
-
-            if (_uiDirty && battery != null && newCharge != null)
-            {
-                UserInterface?.SetState(new ApcBoundInterfaceState(MainBreakerEnabled, extPowerState, newCharge.Value / battery.MaxCharge));
-                _uiDirty = false;
-            }
-        }
-
-        private ApcChargeState CalcChargeState()
-        {
-            if (!_entMan.TryGetComponent(Owner, out BatteryComponent? battery))
-            {
-                return ApcChargeState.Lack;
-            }
-
-            var chargeFraction = battery.CurrentCharge / battery.MaxCharge;
-
-            if (chargeFraction > HighPowerThreshold)
-            {
-                return ApcChargeState.Full;
-            }
+[RegisterComponent]
+[Friend(typeof(ApcSystem))]
+[ComponentProtoName("Apc")]
+public class ApcComponent : BaseApcNetComponent
+{
+    [DataField("onReceiveMessageSound")]
+    public SoundSpecifier OnReceiveMessageSound = new SoundPathSpecifier("/Audio/Machines/machine_switch.ogg");
 
-            var netBattery = _entMan.GetComponent<PowerNetworkBatteryComponent>(Owner);
-            var delta = netBattery.CurrentSupply - netBattery.CurrentReceiving;
+    [ViewVariables]
+    public ApcChargeState LastChargeState;
+    public TimeSpan LastChargeStateTime;
 
-            return delta < 0 ? ApcChargeState.Charging : ApcChargeState.Lack;
-        }
+    // We don't need to cache last charge because we already only run on every charge change
+    // and can thus assume that it will be different.
+    public TimeSpan LastChargeChange;
 
-        private ApcExternalPowerState CalcExtPowerState()
-        {
-            var bat = Battery;
-            if (bat == null)
-                return ApcExternalPowerState.None;
+    [ViewVariables]
+    public ApcExternalPowerState LastExternalState;
+    public TimeSpan LastExternalStateTime;
 
-            var netBat = _entMan.GetComponent<PowerNetworkBatteryComponent>(Owner);
-            if (netBat.CurrentReceiving == 0 && netBat.LoadingNetworkDemand != 0)
-            {
-                return ApcExternalPowerState.None;
-            }
+    [ViewVariables]
+    public bool MainBreakerEnabled = true;
 
-            var delta = netBat.CurrentReceiving - netBat.LoadingNetworkDemand;
-            if (!MathHelper.CloseToPercent(delta, 0, 0.1f) && delta < 0)
-            {
-                return ApcExternalPowerState.Low;
-            }
+    public const float HighPowerThreshold = 0.9f;
+    public const int VisualsChangeDelay = 1;
```

Comment:
Why not just make this a timespan directly, instead of calling `TimeSpawn.FromSeconds()` every time its used?